knowledge-base/adr/0031-resilience-options.md

```diff
@@ -0,0 +1,312 @@
+  maxTimeout?: number; // default Infinity.
+  randomize?: boolean; // default true.
+  onRetry: (e: Error) => {}; // default undefined
+}
+
+// in the resiliece call:
+type ResilieceOptions = {
+  timeout: undefined | number;
+  retry: RetryOptions | { service: RetryOptions; target: RetryOptions };
+  circuitBreaker:
+    | CircuitBreakerOptions
+    | { service: CircuitBreakerOptions; target: CircuitBreakerOptions };
+};
+```
+
+The term `service` indicates the calls to the SAP BTP services and the `target` to the requests to the system in the destination.
+The default situation with 10sec timeout, circuit breaker on and retry switched off would be:
+
+```ts
+defaultResilienceOptions = {
+  timeout: 10000,
+  retry: false,
+  circuitBreaker: true
+};
+```
+
+### Variant A - Opinionated
+
+We pick an implementation and only provide options.
+The API would look like:
+
+```ts
+myApi
+  .getAll()
+  .timeout(20) // deprecate
+  .resilience({...}) // ResilieceOptions
+  .execute({
+      enableCircuitBreaker: true, // deprecate
+      timeout: 10, // deprecate
+      destinationName: 'my-dest'
+  });
+executeHttpRequest({
+      destinationName: 'my-dest'
+    },
+    {
+      resilience: {...} // ResilieceOptions
+    }
+);
+```
+
+Pro:
+
+- Easy to use
+- Defaults non-breaking
+- TypeScript shows options
+
+Contra:
+
+- No custom implementation
+- We have to make assumption on reasonable behavior
+
+### Variant B - Middleware (Default off)
+
+- Users can pass a function taking a function with arguments and returning a `Promise<T>`.
+- For example the input could be the `http-request` function with arguments and the return the promise of `http-client`
+- We provide a sample implementation `resilience()` for easy consumption
+- Assumes all resilience is switched off per default
+
+```ts
+myApi
+  .getAll()
+  .middleware(resilience({ retry: 3 }))
+  .execute({ destinationName: 'my-dest' });
+
+type Middleware<T> = <T>(
+  fn: () => Promise<T>,
+  context?: 'service' | 'target'
+) => Promise<T>;
+```
+
+Pro:
+
+- Easy to use
+- Flexible
+- TypeScript shows options
+
+Contra:
+
+- Extension of existing resilience not possible
+
+### Variant C - Middleware (Named)<a name="variant-c" />
+
+- Similar to variant B
+- Assumes that some resilience is switched on per default and our approach considers this.
+- An optional `id` is passed to the `middleware` method.
+- The `id` will give us the flexibility to add and manage additional middlewares in the future.
+
+```ts
+myApi
+  .getAll()
+  .middleware(resilience({ circuitBreaker: false }), id)
+  .execute({ destinationName: 'my-dest' });
+
+type Middleware<T> = <T>(
+  fn: () => Promise<T>,
+  context?: 'service' | 'target'
+) => Promise<T>;
+```
+
+Use Case A :
+
+- User wants to adjust options of resilience.
+- `id` is omitted and options are passed to middleware call.
+- The default resilience middleware function contains a `id` property with content `sdkResilience`.
```

Comment:
Maybe rename the id with `defaultResilience` or `sdkDefaultResilience` to indicate that it is set by default. (Or if we have some more default middlewares, would it be nice to have a table of default `id`s?) 😄

Reply:
there are two separate arguments the resilience function and the id argument. We could use the function name or add a extra property to the function object.

Reply:
I think function name feels pretty hacky... I would advise against that.

Reply:
yes, I would also use the extra property as indicated in the docs. My last sentence was more a explanation point that the function will hold the information